File: tyama_codeeval59.rb
#!/usr/bin/ruby
$h={'0'=>'0','1'=>'1','2'=>'abc','3'=>'def','4'=>'ghi','5'=>'jkl','6'=>'mno','7'=>'pqrs','8'=>'tuv','9'=>'wxyz'}
$i=0
def d(depth,str,ret)
	if depth==str.length
		print ',' if $i>0
		print ret
		$i+=1
	else
		$h[str[depth,1]].each_byte{|e|
			d(depth+1,str,ret+e.chr)
		}
	end
end
while gets
	$i=0
	d(0,$_.chomp,'')
	puts
end

File: tyama_codeeval59.cpp
#include <iostream>
#include <string>
using namespace std;
string h[]={"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
int i=0;
void D(int depth,string &str,string ret){
	if(depth==str.size()){
		if(i>0)cout<<',';
		cout<<ret;
		i++;
	}else{
		for(int k=0;k<h[str[depth]-'0'].size();k++){
			D(depth+1,str,ret+h[str[depth]-'0'][k]);
		}
	}
}
main(){
	string s;
	for(;cin>>s;){
		i=0;
		D(0,s,string(""));
		cout<<endl;
	}
}